File: ios/Plugin/MicInput.swift
//
//  MicInput.swift
//  Plugin
//
//  Created by Skylabs Technology on 6/11/21.
//  Copyright © 2021 Max Lynch. All rights reserved.
//


import Foundation
import AVFoundation

public class MicInput {

    public let engine: AVAudioEngine = AVAudioEngine()
    public let eq: AVAudioUnitEQ = AVAudioUnitEQ(numberOfBands: 3)

    public var micInput: AVAudioInputNode?
    public let micMixer: AVAudioMixerNode = AVAudioMixerNode()
    public var _parent: Mixer
    public var listenerName: String = ""
    
    public var ioPlayer: AVAudioPlayerNode = AVAudioPlayerNode();
    public var ioBuffer: [AVAudioPCMBuffer] = [];
    public var micInputQueue: DispatchQueue
    public var meterQueue: DispatchQueue
    
    public var inputChannelCount: AVAudioChannelCount = 0;
    public var selectedInputChannel: Int = -1;
    public var toFormat: AVAudioFormat;
    


    // TODO 7/5 : Handle audio session interrupts
    init(parent: Mixer, audioId: String){
        _parent = parent
        //        engine = _parent.engine
        micInputQueue = DispatchQueue(label: "mixerPlugin.micInput.queue.\(audioId)", qos: .userInitiated);
        meterQueue = DispatchQueue(label: "mixerPlugin.micMeter.queue.\(audioId)", qos: .userInitiated);
        toFormat = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: 44100.0, channels: 1, interleaved: true)!
    }
    
    deinit {
        print("We are being disposed")
    }
    
    /**
     * Starts initialization of an Mic input. Configures then starts mic and its listeners
     *
     * @param channelSettings
     */
    // MARK: setupAudio
    public func setupAudio(audioFilePath: NSURL, channelSettings: ChannelSettings) {
        setupEq(with: AVAudioFormat(), channelSettings: channelSettings)
    }
    
    /**
     * Sets up EQ for input
     *
     * @param channelSettings
     */
    // MARK: setupEq
    private func setupEq(with format: AVAudioFormat, channelSettings: ChannelSettings) {
        let bassEq = eq.bands[0]
        bassEq.filterType = .lowShelf
        bassEq.frequency = channelSettings.eqSettings!.bassFrequency!
        bassEq.gain = channelSettings.eqSettings!.bassGain!
        bassEq.bypass = false
        
        let midEq = eq.bands[1]
        midEq.filterType = .parametric
        midEq.frequency = channelSettings.eqSettings!.midFrequency!
        midEq.bandwidth = 1
        midEq.gain = channelSettings.eqSettings!.midGain!
        midEq.bypass = false
        
        let trebleEq = eq.bands[2]
        trebleEq.filterType = .highShelf
        trebleEq.frequency = channelSettings.eqSettings!.trebleFrequency!
        trebleEq.gain = channelSettings.eqSettings!.trebleGain!
        trebleEq.bypass = false
        
        configureEngine(with: format, channelSettings: channelSettings)
    }
    
    /**
     * Completes remaining setup for objects and enables EQ
     *
     * @param channelSettings
     */
    // MARK: configureEngine
    public func configureEngine(with format: AVAudioFormat, channelSettings: ChannelSettings) {
        if (engine.isRunning) {
            engine.stop()
        }
        
        micInput = engine.inputNode
        let micFormat = micInput!.outputFormat(forBus: 0)
        inputChannelCount = micFormat.channelCount
//        toFormat = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: 44100.0, channels: 1, interleaved: true)
        selectedInputChannel = channelSettings.channelNumber!
        
        ioPlayer.volume = 0.8
        micMixer.outputVolume = channelSettings.volume!
        
        engine.attach(eq)
        engine.attach(ioPlayer)
        engine.attach(micMixer)

        
        if (channelSettings.channelListenerName != "") {
            listenerName = channelSettings.channelListenerName!
            micMixer.removeTap(onBus: 0)
            micMixer.installTap(onBus: 0, bufferSize: 1024, format: micMixer.outputFormat(forBus: 0), block: handleMetering)
        }
        
        engine.connect(ioPlayer, to: eq, format: toFormat)
        engine.connect(eq, to: micMixer, format: toFormat)
        engine.connect(micMixer, to: engine.mainMixerNode, format: micMixer.outputFormat(forBus: 0))

        engine.prepare();
        
      do {
        try engine.start()
//        self.ioPlayer.prepare(withFrameCount: 8)
        self.ioPlayer.play();
        micInput!.installTap(onBus: 0, bufferSize: 512, format: micFormat, block: handleInputBuffer)
      } catch {
        print("Error starting the player: \(error.localizedDescription)")
      }
    }
    
    /**
     * Local listener that handles metering 
     * 
     * @param buffer
     * @param time
     */
    // MARK: handleMetering
    public func handleMetering(buffer: AVAudioPCMBuffer, time: AVAudioTime) {
        // https://www.raywenderlich.com/21672160-avaudioengine-tutorial-for-ios-getting-started
        meterQueue.async{
            guard let channelData = buffer.floatChannelData else { return }
          
            let channelDataValue = channelData.pointee

            let channelDataValueArray = stride(
              from: 0,
              to: Int(buffer.frameLength),
              by: buffer.stride)
              .map { channelDataValue[$0] }
          
            let rms = sqrt(channelDataValueArray.map {
              return $0 * $0
            }
            .reduce(0, +) / Float(buffer.frameLength))
          
            let avgPower = 20 * log10(rms)
            let response = avgPower < -80 ? -80 : avgPower

            self._parent.notifyListeners(self.listenerName, data: [ResponseParameters.meterLevel : response])
            
        }
    }

    /**
     * Handles looping over and creating a buffer for mic input
     * 
     * @param buffer
     * @param time
     */
    // MARK: handleInputBuffer
    public func handleInputBuffer(buffer: AVAudioPCMBuffer, time: AVAudioTime) {
        micInputQueue.async {
            // Gets all channels from AVAudioPCMBuffer in an array
            let channels = UnsafeBufferPointer(start: buffer.floatChannelData, count: Int(self.inputChannelCount))
            // Gets data from the channel based on array index
            let ch0Data = NSData(bytes: channels[self.selectedInputChannel], length:Int(buffer.frameCapacity * buffer.format.streamDescription.pointee.mBytesPerFrame))
            let audioFormat = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: 44100.0, channels: 1, interleaved: true)
            let newBuffer = AVAudioPCMBuffer(pcmFormat: audioFormat!, frameCapacity: UInt32(ch0Data.length) / audioFormat!.streamDescription.pointee.mBytesPerFrame)
            newBuffer?.frameLength = newBuffer!.frameCapacity
            let newChannels = UnsafeBufferPointer(start: newBuffer?.floatChannelData, count: Int((newBuffer?.format.channelCount)!))
            ch0Data.getBytes(UnsafeMutableRawPointer(newChannels[0]), length: ch0Data.length)

            self.ioPlayer.scheduleBuffer(newBuffer!, completionHandler: nil)
        }
    }

    // MARK: Player Controls

    /**
     * Not Implemented for MicInput
     *
     * @return
     */
    // MARK: playOrPause
    public func playOrPause() -> String {
        return "not implemented"
    }

    /**
     * Not Implemented for MicInput
     *
     * @return
     */
    // MARK: stop
    public func stop() -> String {
        return "not implemented"
    }
    
    /**
     * Always returns true for MicInput
     *
     * @return
     */
    // MARK: isPlaying
    public func isPlaying() -> Bool {
        return true
    }
    
    /**
     * Changes volume for the mic player
     *
     * @param volume
     */
    // MARK: adjustVolume
    public func adjustVolume(volume: Float) {
        micMixer.outputVolume = volume
//        micInput.volume = volume
    }
    
    /**
     * Returns current volume for mic player
     *
     * @return
     */
    // MARK: getCurrentVolume
    public func getCurrentVolume() -> Float {
        return micMixer.outputVolume
//        return micInput.volume
    }
    
    /**
     * Changes EQ output associated with the mic player
     *
     * @param type
     * @param gain
     * @param freq
     */
    // MARK: adjustEq
    public func adjustEq(type: String, gain: Float, freq: Float) {
        if(eq.bands.count < 1) {
            return
        }
        switch type.lowercased() {
        case "bass":
            let bassEq = eq.bands[0]
            bassEq.gain = gain
            bassEq.frequency = freq == -1 ? bassEq.frequency : freq;
            break;
        case "mid":
            let midEq = eq.bands[1]
            midEq.gain = gain
            midEq.frequency = freq == -1 ? midEq.frequency : freq;
            break;
        case "treble":
            let trebleEq = eq.bands[2]
            trebleEq.gain = gain
            trebleEq.frequency = freq == -1 ? trebleEq.frequency : freq;
            break;
        default:
            print("adjustEq: invalid eq type")
        }
    }
    
    /**
     * Returns current tracked EQ
     *
     * @return
     */
    // MARK: getCurrentEq
    public func getCurrentEq() -> [String: Float] {
        if(eq.bands.count < 1) {
            return [:]
        }
        let bassEq = eq.bands[0]
        let midEq = eq.bands[1]
        let trebleEq = eq.bands[2]
        
        return [ResponseParameters.bassGain: bassEq.gain,
                ResponseParameters.bassFrequency: bassEq.frequency,
                ResponseParameters.midGain: midEq.gain,
                ResponseParameters.midFrequency: midEq.frequency,
                ResponseParameters.trebleGain: trebleEq.gain,
                ResponseParameters.trebleFrequency: trebleEq.frequency]
    }
    
    /** 
     * Not implemented for MicInput
     */
    // MARK: getElapsedTime
    public func getElapsedTime() -> [String: Int] {
        return ["statusCode": 1]
    }
    
    /** 
     * Not implemented for MicInput
     */
    // MARK: getTotalTime
    public func getTotalTime() -> [String: Int] {
        return ["statusCode": 1]
    }
    
    /**
     * Stops mic input temporarily and removes meter notifications and alerts listener.
     *
     * Note: processes will continue running for AudioRecord and AudioTrack in thread.
     * This should only be used temporarily
     */
    // MARK: interrupt
    public func interrupt() {
        micMixer.removeTap(onBus: 0)
        micInput!.removeTap(onBus: 0)
        ioPlayer.stop()
        engine.stop()
    }
    
    /**
     * Resumes mic input and starts meter notifications and alerts listener.
     */
    // MARK: resumeFromInterrupt
    public func resumeFromInterrupt() {
        micInput = engine.inputNode
        if (listenerName != "") {
            micMixer.installTap(onBus: 0, bufferSize: 1024, format: micMixer.outputFormat(forBus: 0), block: handleMetering)
        }
        ioPlayer.play()
        do {
            try engine.start()
            micInput!.installTap(onBus: 0, bufferSize: 512, format: micInput!.outputFormat(forBus: 0), block: handleInputBuffer)
            ioPlayer.play()
        } catch let error {
            print("Error resuming from interrupt with error: \(error)")
        }
    }
    
    /**
     * Destroys object and resets state
     *
     * @return
     */
    // MARK: Destroy
    public func destroy() -> [String : String] {
        micMixer.removeTap(onBus: 0)
        micInput?.removeTap(onBus: 0)
        engine.stop()
        return [ResponseParameters.listenerName: self.listenerName,
                ResponseParameters.elapsedTimeEventName: ""]
    }
}

File: android/src/main/java/com/skylabs/mixer/MicInput.java
package com.skylabs.mixer;

import android.media.AudioAttributes;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioRecord;
import android.media.AudioTrack;
import android.media.MediaRecorder;
import android.media.MicrophoneInfo;
import android.media.audiofx.DynamicsProcessing;
import android.media.audiofx.Visualizer;
import android.util.Log;

import com.getcapacitor.JSObject;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MicInput {
    private Mixer _parent;
    private DynamicsProcessing.Eq eq;
    private DynamicsProcessing dp;
    private double currentVolume;
    private String listenerName = "";
    private int selectedChannel;

    private static final String APP_TAG = "Microphone";
    private static final int mSampleRate = 44100;
    private static final int mFormat = AudioFormat.ENCODING_PCM_16BIT;

    private AudioTrack mAudioOutput;

    private AudioRecord mAudioInput;
    private int mInBufferSize;
    public int mOutBufferSize;
    private static boolean mActive = true;

    private Visualizer visualizer;
    private boolean visualizerState = false;
    private Visualizer.MeasurementPeakRms measurementPeakRms;

    private boolean isInterrupt = false;

    public MicInput(Mixer parent) {
        _parent = parent;
    }

    /**
     * Starts initialization of an Mic input. Configures AudioTrack and AudioRecord, then starts mic and its listeners
     *
     * @param channelSettings
     */
    public void setupAudio(ChannelSettings channelSettings) {
        selectedChannel = channelSettings.channelNumber;
        int mOutChannelFormat = getAudioOutFormatEnum(_parent.foundChannelCount);

        mInBufferSize  = AudioRecord.getMinBufferSize(mSampleRate, AudioFormat.CHANNEL_IN_DEFAULT, mFormat);
        mOutBufferSize = AudioTrack.getMinBufferSize(mSampleRate, mOutChannelFormat, mFormat);

        AudioFormat inAudioFormat;
        if(_parent.foundChannelIndexMask != null){
            inAudioFormat = new AudioFormat.Builder()
                    .setSampleRate(mSampleRate)
                    .setEncoding(mFormat)
                    .setChannelIndexMask(_parent.foundChannelIndexMask)
                    .build();
        }
        else {
            inAudioFormat = new AudioFormat.Builder()
                    .setSampleRate(mSampleRate)
                    .setEncoding(mFormat)
                    .setChannelMask(_parent.foundChannelMask)
                    .build();
        }

        mAudioInput = new AudioRecord.Builder()
                .setAudioFormat(inAudioFormat)
                .setAudioSource(MediaRecorder.AudioSource.MIC)
                .setBufferSizeInBytes(mInBufferSize)
                .build();
        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                                                             .setUsage(AudioAttributes.USAGE_MEDIA)
                                                             .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                                                             .setLegacyStreamType(AudioManager.STREAM_MUSIC)
                                                             .setFlags(AudioAttributes.FLAG_AUDIBILITY_ENFORCED)
                                                             .build();
        AudioFormat outAudioFormat = new AudioFormat.Builder()
                                                 .setSampleRate(mSampleRate)
                                                 .setEncoding(mFormat)
                                                 .setChannelMask(mOutChannelFormat)
                                                 .build();
        mAudioOutput = new AudioTrack.Builder()
                                     .setAudioAttributes(audioAttributes)
                                     .setAudioFormat(outAudioFormat)
                                     .setBufferSizeInBytes(mOutBufferSize)
                                     .setTransferMode(AudioTrack.MODE_STREAM)
                                     //.setEncapsulationMode(AudioTrack.ENCAPSULATION_MODE_ELEMENTARY_STREAM)
                                     .setPerformanceMode(AudioTrack.PERFORMANCE_MODE_LOW_LATENCY)
                                     .setSessionId(_parent.audioManager.generateAudioSessionId())
                                     .build();
        if(_parent.preferredInputDevice != null) {
            mAudioInput.setPreferredDevice(_parent.preferredInputDevice);
        }

//        if(_parent.preferredOutputDevice != null) {
//            mAudioOutput.setPreferredDevice(_parent.preferredOutputDevice);
//        }
        setupEq(channelSettings);
    }

    /**
     * Sets up EQ and attaches it to AudioTrack object
     *
     * @param channelSettings
     */
    private void setupEq(ChannelSettings channelSettings) {
        DynamicsProcessing.EqBand bassEq = new DynamicsProcessing.EqBand(true, (float)channelSettings.eqSettings.bassFrequency, (float)channelSettings.eqSettings.bassGain);
        DynamicsProcessing.EqBand midEq = new DynamicsProcessing.EqBand(true, (float)channelSettings.eqSettings.midFrequency, (float)channelSettings.eqSettings.midGain);
        DynamicsProcessing.EqBand trebleEq = new DynamicsProcessing.EqBand(true, (float)channelSettings.eqSettings.trebleFrequency, (float)channelSettings.eqSettings.trebleGain);
        eq = new DynamicsProcessing.Eq(true, true, 3);
        eq.setBand(0, bassEq);
        eq.setBand(1, midEq);
        eq.setBand(2, trebleEq);
        DynamicsProcessing.Config config = new DynamicsProcessing.Config.Builder(
                DynamicsProcessing.VARIANT_FAVOR_FREQUENCY_RESOLUTION,
                1,
                false, 0,
                false, 0,
                true, 3,
                false
        ).setPreferredFrameDuration(10).build();
        dp = new DynamicsProcessing(0, mAudioOutput.getAudioSessionId(), config);
        dp.setPostEqAllChannelsTo(eq);

        configureEngine(channelSettings);
    }

    /**
     * Completes remaining setup for AudioTrack and AudioRecord objects and enables EQ
     *
     * @param channelSettings
     */
    private void configureEngine(ChannelSettings channelSettings) {
        if (!channelSettings.channelListenerName.isEmpty()) {
            listenerName = channelSettings.channelListenerName;
        }
        currentVolume = channelSettings.volume;
        mAudioOutput.setVolume((float)channelSettings.volume);
        mAudioOutput.setPlaybackRate(mSampleRate);
        dp.setEnabled(true);
        initVisualizerListener();
        List<MicrophoneInfo> mics = null;
        try {
            mics = mAudioInput.getActiveMicrophones();
            Log.i("","");
        } catch (IOException e) {
            e.printStackTrace();
        }
        mAudioInput.addOnRoutingChangedListener(_parent.routingListener, null);
        record();
    }

    /**
     * Not Implemented for MicInput
     *
     * @return
     */
    public String playOrPause() {
        return "not implemented";
    }

    /**
     * Not Implemented for MicInput
     *
     * @return
     */
    public String stop() {
        return "not implemented";
    }

    /**
     * Always returns true for MicInput
     *
     * @return
     */
    public boolean isPlaying() {
        return true;
    }

    /**
     * Changes volume for the AudioTrack
     *
     * @param volume
     */
    public void adjustVolume(double volume) {
        currentVolume = volume;
        mAudioOutput.setVolume((float)volume);
    }

    /**
     * Returns current volume for AudioTrack
     *
     * @return
     */
    public double getCurrentVolume() {
        return currentVolume;
    }

    /**
     * Changes EQ output associated with the AudioTrack
     *
     * @param type
     * @param gain
     * @param freq
     */
    public void adjustEq(String type, double gain, double freq) {
        if (eq.getBandCount() < 1) {
            return;
        }
        if (type.equals("bass")) {
            DynamicsProcessing.EqBand bassEq = eq.getBand(0);
            bassEq.setGain((float) gain);
            bassEq.setCutoffFrequency((float) freq);
            eq.setBand(0, bassEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else if (type.equals("mid")) {
            DynamicsProcessing.EqBand midEq = eq.getBand(1);
            midEq.setGain((float) gain);
            midEq.setCutoffFrequency((float) freq);
            eq.setBand(1, midEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else if (type.equals("treble")) {
            DynamicsProcessing.EqBand trebleEq = eq.getBand(2);
            trebleEq.setGain((float) gain);
            trebleEq.setCutoffFrequency((float) freq);
            eq.setBand(2, trebleEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else {
            System.out.println("adjustEq: invalid eq type");
        }
    }

    /**
     * Returns current tracked EQ
     *
     * @return
     */
    public Map<String, Object> getCurrentEq() {
        Map<String, Object> currentEq = new HashMap<String, Object>();
        currentEq.put(ResponseParameters.bassGain, eq.getBand(0).getGain());
        currentEq.put(ResponseParameters.bassFrequency, eq.getBand(0).getCutoffFrequency());
        currentEq.put(ResponseParameters.midGain, eq.getBand(1).getGain());
        currentEq.put(ResponseParameters.midFrequency, eq.getBand(1).getCutoffFrequency());
        currentEq.put(ResponseParameters.trebleGain, eq.getBand(2).getGain());
        currentEq.put(ResponseParameters.trebleFrequency, eq.getBand(2).getCutoffFrequency());
        return currentEq;
    }

    /**
     * Stops mic input temporarily and removes meter notifications and alerts listener.
     *
     * Note: processes will continue running for AudioRecord and AudioTrack in thread.
     * This should only be used temporarily
     */
    public void interrupt() {
        mAudioOutput.setVolume(0);
        if(visualizerState){
            destroyVisualizerListener();
        }
        JSObject response = new JSObject();
        response.put("handlerType", "ROUTE_DEVICE_DISCONNECTED");

        _parent.notifyPluginListeners(_parent.audioSessionListenerName, response);
    }

    /**
     * Resumes mic input and starts meter notifications and alerts listener.
     */
    public void resumeFromInterrupt() {
        mAudioOutput.setVolume((float)currentVolume);
        if(!visualizerState){
            initVisualizerListener();
        }
        JSObject response = new JSObject();
        response.put("handlerType", "ROUTE_DEVICE_RECONNECTED");

        _parent.notifyPluginListeners(_parent.audioSessionListenerName, response);
    }

    /**
     * Destroys object and resets state
     *
     * @return
     */
    public Map<String, Object> destroy() {
        if(visualizerState) {
            destroyVisualizerListener();
        }
        mActive = false;
        Map<String, Object> response = new HashMap<String, Object>();
        response.put(ResponseParameters.listenerName, listenerName);
        response.put(ResponseParameters.elapsedTimeEventName, "");
        return response;
    }

    /**
     * Starts AudioRecord and AudioTrack and starts metering in its own Thread
     */
    private void record() {
        Thread t = new Thread() {
            public void run() {

                Log.d(APP_TAG, "Entered record loop");

                recordLoop();

                Log.d(APP_TAG, "Record loop finished");
            }

            private void recordLoop() {
                if ( mAudioOutput.getState() != AudioTrack.STATE_INITIALIZED || mAudioInput.getState() != AudioTrack.STATE_INITIALIZED) {
                    Log.d(APP_TAG, "Can't start. Race condition?");
                }
                else {
                    try {
                        try { mAudioOutput.play(); } catch (Exception e) { Log.e(APP_TAG, "Failed to start playback"); return; }
                        try { mAudioInput.startRecording(); } catch (Exception e) { Log.e(APP_TAG,"Failed to start recording"); mAudioOutput.stop(); return; }

                        try {
                            ByteBuffer bytes = ByteBuffer.allocateDirect(mInBufferSize);
                            int o = 0;
                            byte[] b = new byte[mInBufferSize];
                            while(mActive) {
                                o = mAudioInput.read(bytes, (mInBufferSize));
                                bytes.get(b);
                                bytes.rewind();
                                b = removeUnusedChannels(selectedChannel, _parent.foundChannelCount, b);
                                mAudioOutput.write(b, 0, o);
                            }
                            Log.d(APP_TAG, "Finished recording");
                        }
                        catch (Exception e) {
                            Log.d(APP_TAG, "Error while recording, aborting.");
                        }
                        try {
                            if(isInterrupt){
                                mAudioOutput.pause();
                            }
                            else {
                                mAudioOutput.stop();
                            }

                        } catch (Exception e) {
                            Log.e(APP_TAG, "Can't stop playback");
                            mAudioInput.stop();
                            return;
                        }
                        try {
                            if(!isInterrupt){
//                                mAudioInput.stop();
                            }
                        } catch (Exception e) {
                            Log.e(APP_TAG, "Can't stop recording");
                            return;
                        }
                    }
                    catch (Exception e) {
                        Log.d(APP_TAG, "Error somewhere in record loop.");
                    }
                }
            }
            private byte[] removeUnusedChannels(int channelNumber, int totalChannels, byte[] buffer) {
                if(totalChannels == 1 ) return buffer;

                byte[] newBuffer = new byte[buffer.length];

                int offset = 2 * channelNumber;
                int totalChannelBytes = totalChannels * 2;

                for(int i = offset; i < buffer.length; i += totalChannelBytes){
                    newBuffer[i] = buffer[i];
                    newBuffer[i+1] = buffer[i+1];
                }
                return newBuffer;
            }
        };
        t.start();
    }

    /**
     * Tries to determine AudioFormat based on found number of channels
     *
     * @param numberOfChannels
     * @return
     */
    private int getAudioOutFormatEnum(int numberOfChannels){
        switch (numberOfChannels){
            case 1:
                return AudioFormat.CHANNEL_OUT_MONO;
            case 2:
                return AudioFormat.CHANNEL_OUT_STEREO;
            case 4:
                return AudioFormat.CHANNEL_OUT_QUAD;
            case 6:
                return AudioFormat.CHANNEL_OUT_5POINT1;
            case 8:
                return AudioFormat.CHANNEL_OUT_7POINT1_SURROUND;
            default:
                return  AudioFormat.CHANNEL_OUT_DEFAULT;
        }
    }

    /**
     * Starts listener for Audio metering.
     *
     * Note: this should only run when being used as it will run continuously in the background
     *       call destroyVisualizerListener to stop process.
     */
    private void initVisualizerListener() {
        if (listenerName.isEmpty()){
            return;
        }
        if(!visualizerState){
            visualizer = new Visualizer(mAudioOutput.getAudioSessionId());
            visualizer.setScalingMode(Visualizer.SCALING_MODE_AS_PLAYED);
            visualizer.setMeasurementMode(Visualizer.MEASUREMENT_MODE_PEAK_RMS);
            visualizer.setCaptureSize(Visualizer.getCaptureSizeRange()[0]);
            measurementPeakRms = new Visualizer.MeasurementPeakRms();
            visualizer.setDataCaptureListener(new Visualizer.OnDataCaptureListener() {
                @Override
                public void onWaveFormDataCapture(Visualizer vis, byte[] bytes, int i) {
                    if(visualizerState){
                        visualizer.getMeasurementPeakRms(measurementPeakRms);
                        double measurement = (double)measurementPeakRms.mRms;
                        measurement = (measurement / 100) * (1 / currentVolume);
                        double response = measurement < -80 ? -80 : measurement;
                        JSObject data = new JSObject();
                        data.put(ResponseParameters.meterLevel, response);
                        _parent.notifyPluginListeners(listenerName, data);
                    }
                }

                @Override
                public void onFftDataCapture(Visualizer visualizer, byte[] bytes, int i) {
                    // Log.i("FFT Byte Array: ", String.valueOf(bytes[0]));
                }
            }, Visualizer.getMaxCaptureRate(), true, false);
            visualizer.setEnabled(true);
            visualizerState = true;
        }
    }

    /**
     * Stop listener for audio metering.
     */
    private void destroyVisualizerListener() {
        if (listenerName.isEmpty()){
            return;
        }
        if (visualizerState) {
            visualizer.setDataCaptureListener(null, 0, false, false);
            visualizer.setEnabled(false);
            visualizerState = false;
            visualizer.release();
        }
    }
}
